packages/@lwc/shared/src/html-attributes.ts

```diff
@@ -178,3 +189,35 @@ export function kebabCaseToCamelCase(attrName: string): string {
 
     return result;
 }
+
+export function normalizeClass(value: unknown): string | undefined {
+    if (isUndefined(value) || isNull(value)) {
+        // Returning undefined here improves initial render cost, because the old vnode's class will be considered
+        // undefined in the `patchClassAttribute` routine, so `oldClass === newClass` will be true so we return early
+        return undefined;
+    }
+
+    let res = '';
+
+    if (isString(value)) {
+        res = value;
+    } else if (isArray(value)) {
+        for (let i = 0; i < value.length; i++) {
+            const normalized = normalizeClass(value[i]);
+            if (normalized) {
+                res += normalized + ' ';
+            }
+        }
+    } else if (isObject(value) && !isNull(value)) {
+        // Iterate own enumerable keys of the object
+        const _keys = keys(value);
+        for (let i = 0; i < _keys.length; i += 1) {
+            const key = _keys[i];
+            if ((value as Record<string, unknown>)[key]) {
+                res += key + ' ';
+            }
+        }
+    }
+
+    return StringTrim.call(res);
+}
```

Comment:
Running something recursive server-side is a little bit scary, but I don't think it's possible to hit an infinite loop unless doing something really weird.

Reply:
Yeah we are typically not too concerned about this, because the customer (in our case component authors) can only hose themselves. They are indeed running on our infra, but there are always ways they can hose themselves (e.g. having massively nested `<template for:each>` in an HTML file, or massively nested `if` blocks in a JS file, or tons of `@media` nested inside themselves etc. – these can all hose Rollup/Babel/PostCSS/Parse5/etc).

Reply:
Also: could you please move the function-level comment here and move this function to its own `normalize-class.ts` file? I don't think it fits well with `html-attributes.ts` since it's its own LWC-specific class-munging logic.